action/protocol/rewarding/protocol.go

```diff
@@ -396,29 +419,30 @@ func (p *Protocol) settleAction(
 			return nil, err
 		}
 	}
-	gasFee := big.NewInt(0).Mul(actionCtx.GasPrice, big.NewInt(0).SetUint64(actionCtx.IntrinsicGas))
-	depositLog, err := DepositGas(ctx, sm, gasFee)
-	if err != nil {
-		return nil, err
-	}
-	if depositLog != nil {
-		tLogs = append(tLogs, depositLog)
-	}
-	if err := p.increaseNonce(sm, actionCtx.Caller, actionCtx.Nonce); err != nil {
-		return nil, err
+	if !isSystemAction || !protocol.MustGetFeatureCtx(ctx).SkipUpdateForSystemAction {
+		gasFee := big.NewInt(0).Mul(actionCtx.GasPrice, big.NewInt(0).SetUint64(actionCtx.IntrinsicGas))
+		depositLog, err := DepositGas(ctx, sm, gasFee)
+		if err != nil {
+			return nil, err
+		}
+		if depositLog != nil {
+			tLogs = append(tLogs, depositLog)
+		}
+		if err := p.increaseNonce(sm, actionCtx.Caller, actionCtx.Nonce, isSystemAction); err != nil {
+			return nil, err
+		}
 	}
 	return p.createReceipt(status, blkCtx.BlockHeight, actionCtx.ActionHash, actionCtx.IntrinsicGas, logs, tLogs...), nil
 }
 
-func (p *Protocol) increaseNonce(sm protocol.StateManager, addr address.Address, nonce uint64) error {
+func (p *Protocol) increaseNonce(sm protocol.StateManager, addr address.Address, nonce uint64, isSystemAction bool) error {
```

Comment:
we don't need `isSystemAction`? for systemAction, its caller won't change and this func will be a no-op?

Reply:
the current implementation is misleading. it will ignore all cases where `nonce < acc.PendingNonce()`. `nonce < acc.PendingNonce()` is not identical to `nonce == 0 && isSystemAction`.